Zoltan2: MultiJagged causing Epetra-based Panzer MiniEM RefMaxwell test to fail

## Bug Report
@trilinos/zoltan2 
@trilinos/panzer 

### Description
In a fresh build from develop with tests and examples enabled, Panzer's ``PanzerMiniEM_MiniEM-BlockPrec_RefMaxwell_Epetra_MPI_4`` is failing. This is in panzer/mini-em/example/BlockPrec. It's really an example (not a normal test), but this is probably worth looking into. Changing the MueLu repartition algorithm in ``solverMueLuRefMaxwellEpetra.xml`` from multijagged to rcb makes it pass.

### Steps to Reproduce
1. SHA1:  8565fc6 (current develop as of 12/10 afternoon)
2. [Configure script](https://gist.github.com/brian-kelley/91addd5352420460bd365f034efd9077)
3. Error output I get:
```
******* WARNING *******
Level::Set: unable to store "AP reuse data" generated by factory RAPFactory[15](0x69009f0) on level 1, as it has not been requested and no keep flags were set for it

******* WARNING *******
Level::Set: unable to store "RAP reuse data" generated by factory RAPFactory[15](0x69009f0) on level 1, as it has not been requested and no keep flags were set for it
RefMaxwell (2,2): Build (MueLu::RepartitionFactory)
RefMaxwell (2,2): Build (MueLu::Zoltan2Interface)
Zoltan2 parameters:
----------
algorithm = multijagged
partitioning_approach = partition   [unused]
mj_premigration_option = 1   [unused]
num_global_parts = 3   [unused]
----------
RefMaxwell (2,2): Zoltan2 multijagged
  Teko: "buildInverse" could not construct the inverse operator using "Thyra::MueLuRefMaxwellPreconditionerFactory"
  
  *** THROWN EXCEPTION ***
  getIDsKokkosView called but not on Tpetra!
  ************************
terminate called after throwing an instance of 'terminate called after throwing an instance of 'std::exception'
std::exception'
  what():  std::exception  what():  std::exception

[s1042556:101025] *** Process received signal ***
[s1042556:101026] *** Process received signal ***
  Teko: "buildInverse" could not construct the inverse operator using "Teko::AutoClone<mini_em::FullMaxwellPreconditionerFactory, mini_em::FullMaxwellPreconditionerFactory>"
  
  *** THROWN EXCEPTION ***
  std::exception
  ************************
[s1042556:101026] Signal: Aborted (6)
```

@brian-kelley Got further but still having issues. I'm trying to login to rhel7T first.

I'm failing now with:
```
[ 10%] Linking CXX executable Epetra_BlockMap_test.exe
../../src/libepetra.so.12.17: undefined reference to `dggsvd_'
../../src/libepetra.so.12.17: undefined reference to `sggsvd_'
```

This is my config for Lapack:
```
Processing enabled TPL: LAPACK (enabled by TeuchosNumerics, disable with -DTPL_ENABLE_LAPACK=OFF)
-- LAPACK_LIBRARY_NAMES='lapack lapack_win32'
-- Searching for libs in LAPACK_LIBRARY_DIRS='/ascldap/users/projects/ppc64le-pwr8/netlib/3.8.0/gcc/7.2.0/lib'
-- Searching for a lib in the set "lapack lapack_win32":
--   Searching for lib 'lapack' ...
--     Found lib '/ascldap/users/projects/ppc64le-pwr8/netlib/3.8.0/gcc/7.2.0/lib/liblapack.a'
-- TPL_LAPACK_LIBRARIES='/ascldap/users/projects/ppc64le-pwr8/netlib/3.8.0/gcc/7.2.0/lib/liblapack.a'
```

My procedure is:
Fresh login to white
bsub -Is -n 16 -q rhel7T bash
source /ascldap/users/medemes/trilinos/cmake/std/atdm/load-env.sh gnu
Then run above config in a fresh empty build folder.
make NP=16

--- a/packages/zoltan2/src/input/Zoltan2_XpetraMultiVectorAdapter.hpp
+++ b/packages/zoltan2/src/input/Zoltan2_XpetraMultiVectorAdapter.hpp
@@ -145,8 +145,17 @@ public:
         dynamic_cast<const xt_mvector_t *>(vector_.get());
       ids = tvector->getTpetra_MultiVector()->getMap()->getMyGlobalIndices();
     }
+    else if (map_->lib() == Xpetra::UseEpetra) {
+#if defined(HAVE_ZOLTAN2_EPETRA) && defined(HAVE_XPETRA_EPETRA)
+      // this will call getIDsView to get raw ptr and create a view from it
+      return BaseAdapter<User>::getIDsKokkosView(ids);
+#else
+      throw std::logic_error("Epetra requested, but Trilinos is not "
+                           "built with Epetra");
+#endif
+    }
     else {
-      throw std::logic_error("getIDsKokkosView called but not on Tpetra!");
+      throw std::logic_error("getIDsKokkosView called but not on Tpetra or Epetra!");
     }
   }
 
@@ -322,14 +331,33 @@ template <typename User>
     // coordinates in MJ are LayoutLeft since Tpetra Multivector gives LayoutLeft
     Kokkos::View<scalar_t **, Kokkos::LayoutLeft, typename node_t::device_type> view2d =
       tvector->getTpetra_MultiVector()->template getLocalView<node_t>();
-
     elements = view2d;
-
-    // TODO: Delete later ... keeping in case we end up switching back
-    // elements = Kokkos::subview(view2d, Kokkos::ALL, idx);
+  }
+  else if (map_->lib() == Xpetra::UseEpetra){
+#if defined(HAVE_ZOLTAN2_EPETRA) && defined(HAVE_XPETRA_EPETRA)
+    typedef Xpetra::EpetraMultiVectorT<gno_t,node_t> xe_mvector_t;
+    const xe_mvector_t *evector =
+      dynamic_cast<const xe_mvector_t *>(vector_.get());
+    elements =
+      Kokkos::View<scalar_t **, Kokkos::LayoutLeft, typename node_t::device_type>
+      ("elements", evector->getLocalLength(), evector->getNumVectors());
+    if(evector->getLocalLength() > 0) {
+      for(size_t idx = 0; idx < evector->getNumVectors(); ++idx) {
+        const scalar_t * ptr;
+        int stride;
+        getEntriesView(ptr, stride, idx);
+        for(size_t n = 0; n < evector->getLocalLength(); ++n) {
+          elements(n, idx) = ptr[n];
+        }
+      }
+    }
+#else
+    throw std::logic_error("Epetra requested, but Trilinos is not "
+                           "built with Epetra");
+#endif
   }
   else {
-    throw std::logic_error("getEntriesKokkosView called but not using Tpetra!");
+    throw std::logic_error("getEntriesKokkosView called but not using Tpetra or Epetra!");
   }
 }
 

--- a/packages/zoltan2/test/partition/CMakeLists.txt
+++ b/packages/zoltan2/test/partition/CMakeLists.txt
@@ -636,3 +636,21 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
 
 ##########################################
 
+IF (NOT Tpetra_ENABLE_CUDA)
+IF (${PACKAGE_NAME}_ENABLE_Epetra)
+IF (Tpetra_INST_INT_INT)
+
+# Exercises the Epetra path for the XpetraMultiVector
+# interfacing with the Multijagged algorithm.
+TRIBITS_ADD_EXECUTABLE_AND_TEST(
+    mj_epetra
+    SOURCES mj_epetra.cpp
+    COMM serial mpi
+    PASS_REGULAR_EXPRESSION "PASS"
+    FAIL_REGULAR_EXPRESSION "FAIL"
+)
+
+ENDIF()
+ENDIF()
+ENDIF()
+

--- a/packages/zoltan2/test/partition/mj_epetra.cpp
+++ b/packages/zoltan2/test/partition/mj_epetra.cpp
@@ -1,0 +1,134 @@
+// @HEADER
+//
+// ***********************************************************************
+//
+//   Zoltan2: A package of combinatorial algorithms for scientific computing
+//                  Copyright 2012 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Karen Devine      (kddevin@sandia.gov)
+//                    Erik Boman        (egboman@sandia.gov)
+//                    Siva Rajamanickam (srajama@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+
+/*! \file mj_epetra
+    \brief Protects the epetra acess to Zoltan2_XpetraMultiVectorAdapter.hpp.
+           Tests via a multijagged partitioning.
+*/
+
+#include <vector>
+#include <numeric>
+#include "Epetra_ConfigDefs.h"
+#ifdef HAVE_MPI
+#include <Epetra_MpiComm.h>
+#else
+#include <Epetra_SerialComm.h>
+#endif
+#include <Epetra_MultiVector.h>
+#include <Zoltan2_PartitioningSolution.hpp>
+#include <Zoltan2_PartitioningProblem.hpp>
+#include <Zoltan2_XpetraMultiVectorAdapter.hpp>
+#include <Zoltan2_InputTraits.hpp>
+
+int main(int narg, char *arg[])
+{
+  Tpetra::ScopeGuard scope(&narg, &arg);
+  const Teuchos::RCP<const Teuchos::Comm<int> > comm = Tpetra::getDefaultComm();
+
+  typedef int gid_t;
+
+  const int N = 100; // num coords
+  const int num_parts = 5;
+  const int dim = 3;
+
+  int rank = comm->getRank();
+
+  // params
+  Teuchos::ParameterList params("test params");
+  params.set("algorithm", "multijagged");
+  params.set("num_global_parts", num_parts);
+
+  // create gids
+  std::vector<gid_t> global_ids(N);
+  std::iota(global_ids.begin(), global_ids.end(), rank * N);
+
+#ifdef HAVE_MPI
+  Epetra_MpiComm epetra_comm(MPI_COMM_WORLD);
+#else
+  Epetra_SerialComm epetra_comm;
+#endif
+
+  Epetra_BlockMap map(-1, N, 1, 0, epetra_comm);
+
+  typedef Epetra_MultiVector mv_t;
+  Teuchos::RCP<mv_t> mv = Teuchos::rcp(new mv_t(map, dim)); // all 0's
+
+  std::vector<int> stride;
+  std::vector<const double *> weights;
+  Teuchos::RCP<const mv_t> cmv = Teuchos::rcp_const_cast<const mv_t>(mv);
+
+  typedef Zoltan2::XpetraMultiVectorAdapter<mv_t> inputAdapter_t;
+
+  Teuchos::RCP<inputAdapter_t> ia =
+    Teuchos::rcp(new inputAdapter_t(cmv, weights, stride));
+
+  Teuchos::RCP<Zoltan2::PartitioningProblem<inputAdapter_t>> problem =
+    Teuchos::rcp(new Zoltan2::PartitioningProblem<inputAdapter_t>(ia.get(), &params));
+
+  problem->solve();
+
+  typedef Zoltan2::EvaluatePartition<inputAdapter_t> quality_t;
+  Teuchos::RCP<quality_t> metricObject = Teuchos::rcp(new quality_t(
+    ia.get(), &params, comm, &problem->getSolution()));
+
+  int err = 0;
+  if (comm->getRank() == 0) {
+    metricObject->printMetrics(std::cout);
+    double imb = metricObject->getObjectCountImbalance();
+    if (imb <= 1.01)
+      std::cout << "balance satisfied " << imb << std::endl;
+    else {
+      std::cout << "balance failed " << imb << std::endl;
+      err++;
+    }
+  }
+
+  if(rank == 0 && err == 0) {
+    std::cout << "PASS" << std::endl;
+  }
+
+  return 0;
+}
+

